crates/evm/src/execute.rs

```diff
@@ -161,6 +162,152 @@ pub trait BlockExecutorProvider: Send + Sync + Clone + Unpin + 'static {
         DB: Database<Error: Into<ProviderError> + Display>;
 }
 
+/// A factory for creating block execution strategies.
+pub trait BlockExecutionStrategyFactory<DB>
+where
+    DB: Database,
+{
+    /// The specific [`BlockExecutionStrategy`] type this factory produces.
+    type Strategy: BlockExecutionStrategy<DB>;
+
+    /// The error type returned by this factory and its produced strategies.
+    type Error: From<ProviderError>
+        + From<<Self::Strategy as BlockExecutionStrategy<DB>>::Error>
+        + core::error::Error;
+
+    /// Creates a new block execution strategy instance.
+    fn create(
+        &self,
+        block: &BlockWithSenders,
+        total_difficulty: U256,
+    ) -> Result<Self::Strategy, Self::Error>;
+}
+
+/// Defines the strategy for executing a single block.
+pub trait BlockExecutionStrategy<DB> {
+    /// The error type returned by this strategy's methods.
+    type Error: From<ProviderError> + core::error::Error;
+
+    /// Configures the environment before start applying any changes.
+    fn configure_environment(&mut self) -> Result<(), Self::Error>;
+
+    /// Applies any necessary changes before executing the block's transactions.
+    fn apply_pre_execution_changes(&mut self) -> Result<(), Self::Error>;
+
+    /// Executes all transactions in the block.
+    fn execute_transactions(
+        &mut self,
+        block: &BlockWithSenders,
+    ) -> Result<(Vec<Receipt>, u64), Self::Error>;
+
+    /// Applies any necessary changes after executing the block's transactions.
+    fn apply_post_execution_changes(&mut self) -> Result<Vec<Request>, Self::Error>;
+
+    /// Returns a reference to the current state.
+    fn state_ref(&self) -> &State<DB>;
+
+    /// Sets a hook to be called after each state change during execution.
+    fn set_state_hook<F: OnStateHook + 'static>(&mut self, hook: Option<F>);
+
+    /// Consumes the strategy and returns the final bundle state.
+    fn finish(self) -> BundleState;
+}
+
+/// A generic block executor that uses a [`BlockExecutionStrategyFactory`] to
+/// create and run block execution strategies.
+#[allow(missing_debug_implementations)]
+pub struct GenericBlockExecutor<F, DB>
+where
+    F: BlockExecutionStrategyFactory<DB>,
+    DB: Database,
+{
+    strategy_factory: F,
+    _phanttom: PhantomData<DB>,
+}
```

Comment:
I think we're missing something here, if we look at how the equivalent is currently initialized:

https://github.com/paradigmxyz/reth/blob/84370b81d7a2e2cb90df1a8d5734469e9b89f5fc/crates/ethereum/evm/src/execute.rs#L60-L74

which gits the DB as input.

I wonder if this type is actually the equivalent of `EthExecutorProvider` that uses a strategy_factory to create a new instance of a strategy wrapped in a generic executor

Reply:
absolutely, in this initial pass i was not including all the details for making it fit with the current setup, will do it now